api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java

```diff
@@ -0,0 +1,617 @@
+        reader.useBoolean("italic", value -> style.decoration(TextDecoration.ITALIC, TextDecoration.State.byBoolean(value)));
+        reader.useBoolean("underlined", value -> style.decoration(TextDecoration.UNDERLINED, TextDecoration.State.byBoolean(value)));
+        reader.useBoolean("strikethrough", value -> style.decoration(TextDecoration.STRIKETHROUGH, TextDecoration.State.byBoolean(value)));
+        reader.useBoolean("obfuscated", value -> style.decoration(TextDecoration.OBFUSCATED, TextDecoration.State.byBoolean(value)));
+        reader.useUTF("insertion", style::insertion);
+
+        NBTReader clickEvent = reader.child("clickEvent");
+        if (clickEvent != null) {
+            style.clickEvent(ClickEvent.clickEvent(
+                    clickEvent.readUTF("action", ClickEvent.Action.NAMES::value),
+                    clickEvent.readUTF("value", Function.identity())
+            ));
+        }
+
+        NBTReader hoverEvent = reader.child("hoverEvent");
+        if (hoverEvent != null) {
+            HoverEvent.Action action = hoverEvent.readUTF("action", HoverEvent.Action.NAMES::value);
+            if (action.equals(HoverEvent.Action.SHOW_TEXT)) {
+                style.hoverEvent(HoverEvent.showText(hoverEvent.read("contents", this::deserialize)));
+            } else if (action.equals(HoverEvent.Action.SHOW_ITEM)) {
+                if (hoverEvent.type("contents") == NBTType.STRING) {
+                    style.hoverEvent(HoverEvent.showItem(hoverEvent.readUTF("contents", Key::key), 1));
+                } else {
+                    NBTReader child = hoverEvent.child("contents");
+                    style.hoverEvent(HoverEvent.showItem(
+                            child.readUTF("id", Key::key),
+                            child.readInt("count", Function.identity()),
+                            child.readUTF("tag", BinaryTagHolder::binaryTagHolder)
+                    ));
+                }
+            } else if (action.equals(HoverEvent.Action.SHOW_ENTITY)) {
+                NBTReader child = hoverEvent.child("contents");
+                style.hoverEvent(HoverEvent.showEntity(
+                        child.readUTF("type", Key::key),
+                        child.readIntArray("id", this::deserializeUUID),
+                        child.read("name", this::deserialize)
+                ));
+            }
+        }
+
+        return style.build();
+    }
+
+    public @NotNull NBTCompound serializeStyle(Style style) {
+        if (style.isEmpty()) return new NBTCompound();
+
+        NBTWriter writer = new NBTWriter(new NBTCompound());
+
+        Key font = style.font();
+        if (font != null) writer.writeUTF("font", font.asString());
+
+        TextColor color = style.color();
+        if (color != null) writer.writeUTF("color", serializeColor(color));
+
+        TextDecoration.State bold = style.decoration(TextDecoration.BOLD);
```

Comment:
This whole TextDecoration block could now be replaced by just iterating all existing TextDecorations, as this no longer is one giant switch

Reply:
Thanks for pointing those out